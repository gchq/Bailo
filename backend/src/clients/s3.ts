import {
  CreateBucketCommand,
  CreateBucketRequest,
  GetObjectCommand,
  GetObjectRequest,
  HeadBucketCommand,
  HeadBucketRequest,
  HeadObjectRequest,
  S3Client,
  S3ServiceException,
} from '@aws-sdk/client-s3'
import { Upload } from '@aws-sdk/lib-storage'
import { NodeHttpHandler } from '@smithy/node-http-handler'
import prettyBytes from 'pretty-bytes'
import { PassThrough, Readable } from 'stream'

import { getHttpsAgent } from '../services/http.js'
import log from '../services/log.js'
import config from '../utils/config.js'

export async function getS3Client() {
  return new S3Client({
    ...(config.s3.credentials.accessKeyId &&
      config.s3.credentials.secretAccessKey && { credentials: { ...config.s3.credentials } }),
    endpoint: config.s3.endpoint,
    region: config.s3.region,
    forcePathStyle: config.s3.forcePathStyle,
    requestHandler: new NodeHttpHandler({
      httpsAgent: getHttpsAgent({ rejectUnauthorized: config.s3.rejectUnauthorized }),
    }),
  })
}

export async function putObjectStream(
  bucket: string,
  key: string,
  body: PassThrough | Readable,
  metadata?: Record<string, string>,
) {
  const upload = new Upload({
    client: await getS3Client(),
    params: { Bucket: bucket, Key: key, Body: body, Metadata: metadata },
    queueSize: 4,
    partSize: 1024 * 1024 * 64,
    leavePartsOnError: false,
  })

  let fileSize = 0
  upload.on('httpUploadProgress', (progress) => {
    log.debug(
      {
        ...progress,
        ...(progress.loaded && { loaded: prettyBytes(progress.loaded), loadedBytes: progress.loaded }),
        ...(progress.total && { total: prettyBytes(progress.total), totalBytes: progress.total }),
      },
      'Object upload is in progress',
    )
    if (progress.loaded) {
      fileSize = progress.loaded
    }
  })

  const s3Response = await upload.done()
  log.debug(s3Response, 'Object upload complete')

  return {
    fileSize,
  }
}

export async function getObjectStream(bucket: string, key: string, range?: { start: number; end: number }) {
  const client = await getS3Client()

  const input: GetObjectRequest = {
    Bucket: bucket,
    Key: key,
  }

  if (range) {
    input.Range = `bytes=${range.start}-${range.end}`
  }

  const command = new GetObjectCommand(input)
  const response = await client.send(command)

  return response
}

export async function headObject(bucket: string, key: string) {
  const client = await getS3Client()

  const input: HeadObjectRequest = {
    Bucket: bucket,
    Key: key,
  }

  const command = new GetObjectCommand(input)
  const response = await client.send(command)

  return response
}

export async function objectExists(bucket: string, key: string) {
  try {
    log.info({ bucket, key }, `Searching for ${key} in ${bucket}`)
    await headObject(bucket, key)
    return true
  } catch (error) {
    if (isS3ServiceException(error) && error.$metadata.httpStatusCode === 404) {
      log.info({ bucket, key }, `Failed to find ${key} in ${bucket}`)
      return false
    } else {
      throw error
    }
  }
}

export async function headBucket(bucket: string) {
  const client = await getS3Client()

  const input: HeadBucketRequest = {
    Bucket: bucket,
  }

  const command = new HeadBucketCommand(input)
  const response = await client.send(command)

  return response
}

export async function createBucket(bucket: string) {
  const client = await getS3Client()

  const input: CreateBucketRequest = {
    Bucket: bucket,
  }

  const command = new CreateBucketCommand(input)
  const response = await client.send(command)

  return response
}

export async function ensureBucketExists(bucket: string) {
  try {
    log.info({ bucket }, `Ensuring ${bucket} exists`)
    await headBucket(bucket)
    log.info({ bucket }, `Bucket ${bucket} already exists`)
  } catch (error) {
    if (isS3ServiceException(error) && error.$metadata.httpStatusCode === 404) {
      log.info({ bucket }, `Bucket does not exist, creating ${bucket}`)
      return createBucket(bucket)
    }
    throw error
  }
}

const isS3ServiceException = (value: unknown): value is S3ServiceException => {
  if (typeof value !== 'object' || value === null) {
    return false
  }
  if (!('name' in value) || !('message' in value) || !('$fault' in value) || !('$metadata' in value)) {
    return false
  }
  return true
}
