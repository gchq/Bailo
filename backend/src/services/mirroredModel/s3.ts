import { Readable } from 'node:stream'

import { sign } from '../../clients/kms.js'
import { getObjectStream, putObjectStream } from '../../clients/s3.js'
import config from '../../utils/config.js'
import { InternalError } from '../../utils/error.js'
import log from '../log.js'
import { generateDigest } from './mirroredModel.js'

export async function uploadToS3(fileName: string, stream: Readable, logData?: Record<string, unknown>) {
  if (config.modelMirror.export.kmsSignature.enabled) {
    log.debug(logData, 'Using signatures. Uploading to temporary S3 location first.')
    try {
      await uploadToTemporaryS3Location(fileName, stream, logData)
      await copyToExportBucketWithSignatures(fileName, logData)
    } catch (error) {
      log.error({ error, ...logData }, 'Failed to upload export to export location with signatures')
    }
  } else {
    log.debug(logData, 'Signatures not enabled. Uploading to export S3 location.')
    try {
      await uploadToExportS3Location(fileName, stream, logData)
    } catch (error) {
      log.error({ error, ...logData }, 'Failed to upload export to export location without signatures')
    }
  }
}

async function copyToExportBucketWithSignatures(fileName: string, logData?: Record<string, unknown>) {
  let signatures = {}
  log.debug(logData, 'Getting stream from S3 to generate signatures.')
  const streamForDigest = await getObjectFromTemporaryS3Location(fileName, logData)
  const messageDigest = await generateDigest(streamForDigest)
  log.debug(logData, 'Generating signatures.')
  try {
    signatures = await sign(messageDigest)
  } catch (e) {
    log.error(logData, 'Error generating signature for export.')
    throw e
  }
  log.debug(logData, 'Successfully generated signatures')
  log.debug(logData, 'Getting stream from S3 to upload to export location.')
  const streamToCopy = await getObjectFromTemporaryS3Location(fileName, logData)
  await uploadToExportS3Location(fileName, streamToCopy, logData, signatures)
}

async function uploadToTemporaryS3Location(
  fileName: string,
  stream: Readable,
  logData?: Record<string, unknown>,
  metadata?: Record<string, string>,
) {
  const bucket = config.s3.buckets.uploads
  const object = `exportQueue/${fileName}`
  try {
    await putObjectStream(object, stream, undefined, metadata)
    log.debug(
      {
        bucket,
        object,
        ...(metadata && { metadata }),
        ...logData,
      },
      'Successfully uploaded export to temporary S3 location.',
    )
  } catch (error) {
    throw InternalError('Failed to export to temporary S3 location.', {
      bucket,
      object,
      ...(metadata && { metadata }),
      error,
      ...logData,
    })
  }
}

async function getObjectFromTemporaryS3Location(fileName: string, logData?: Record<string, unknown>) {
  const bucket = config.s3.buckets.uploads
  const object = `exportQueue/${fileName}`
  try {
    const stream = (await getObjectStream(object, bucket)).Body as Readable
    log.debug(
      {
        bucket,
        object,
        ...logData,
      },
      'Successfully retrieved stream from temporary S3 location.',
    )
    return stream
  } catch (error) {
    throw InternalError('Failed to retrieve stream from temporary S3 location.', {
      bucket,
      object,
      error,
      ...logData,
    })
  }
}

export async function getObjectFromExportS3Location(object: string, logData?: Record<string, unknown>) {
  const bucket = config.modelMirror.export.bucket
  try {
    const stream = (await getObjectStream(object, bucket)).Body as Readable
    log.debug(
      {
        bucket,
        object,
        ...logData,
      },
      'Successfully retrieved stream from temporary S3 location.',
    )
    return stream
  } catch (error) {
    log.error(
      {
        bucket,
        object,
        error,
        ...logData,
      },
      'Failed to retrieve stream from temporary S3 location.',
    )
    throw error
  }
}

async function uploadToExportS3Location(
  object: string,
  stream: Readable,
  logData?: Record<string, unknown>,
  metadata?: Record<string, string>,
) {
  const bucket = config.modelMirror.export.bucket
  try {
    await putObjectStream(object, stream, bucket, metadata)
    log.debug(
      {
        bucket,
        object,
        ...logData,
      },
      'Successfully uploaded export to export S3 location.',
    )
  } catch (error) {
    log.error(
      {
        bucket,
        object,
        error,
        ...logData,
      },
      'Failed to export to export S3 location.',
    )
    throw InternalError('Failed to export to export S3 location.', {
      bucket,
      object,
      error,
      ...logData,
    })
  }
}
