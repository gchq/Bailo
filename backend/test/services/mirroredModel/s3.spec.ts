import { PassThrough, Readable } from 'node:stream'

import { describe, expect, test, vi } from 'vitest'

import { getObjectFromExportS3Location, uploadToS3 } from '../../../src/services/mirroredModel/s3.js'

const configMock = vi.hoisted(
  () =>
    ({
      ui: {
        modelMirror: {
          import: {
            enabled: true,
          },
          export: {
            enabled: true,
          },
        },
      },
      s3: { buckets: { uploads: 'test' } },
      modelMirror: {
        export: {
          maxSize: 100,
          kmsSignature: {
            enabled: true,
          },
        },
      },
      registry: {
        connection: {
          internal: 'https://localhost:5000',
        },
      },
      connectors: {
        audit: {
          kind: 'silly',
        },
      },
    }) as any,
)
vi.mock('../../../src/utils/config.js', () => ({
  __esModule: true,
  default: configMock,
}))

const logMock = vi.hoisted(() => ({
  info: vi.fn(),
  debug: vi.fn(),
  warn: vi.fn(),
  error: vi.fn(),
}))
vi.mock('../../../src/services/log.js', async () => ({
  default: logMock,
}))

const s3Mocks = vi.hoisted(() => ({
  putObjectStream: vi.fn(() => Promise.resolve({ fileSize: 100 })),
  getObjectStream: vi.fn(() => Promise.resolve({ Body: new PassThrough() })),
  objectExists: vi.fn(() => Promise.resolve(true)),
}))
vi.mock('../../../src/clients/s3.js', () => s3Mocks)

const schemaMock = vi.hoisted(() => ({
  generateDigest: vi.fn(),
}))
vi.mock('../../../src/services/mirroredModel/mirroredModel.js', async () => schemaMock)

const kmsMocks = vi.hoisted(() => ({
  sign: vi.fn(),
}))
vi.mock('../../../src/clients/kms.js', () => kmsMocks)

describe('services > mirroredModel > s3', () => {
  test('uploadToS3 > single S3 upload when kms not enabled', async () => {
    vi.spyOn(configMock, 'modelMirror', 'get').mockReturnValue({
      enabled: true,
      export: {
        maxSize: 10,
        kmsSignature: {
          enabled: false,
        },
      },
    })
    await uploadToS3('', {} as unknown as Readable, {} as any)

    expect(s3Mocks.putObjectStream).toHaveBeenCalledTimes(1)
    expect(s3Mocks.getObjectStream).not.toHaveBeenCalled()
  })

  test('uploadToS3 > Double S3 upload when kms enabled', async () => {
    vi.spyOn(configMock, 'modelMirror', 'get').mockReturnValue({
      enabled: true,
      export: {
        maxSize: 10,
        kmsSignature: {
          enabled: true,
        },
      },
    })
    await uploadToS3('', {} as unknown as Readable, {} as any)

    expect(s3Mocks.putObjectStream).toHaveBeenCalledTimes(2)
    expect(s3Mocks.getObjectStream).toHaveBeenCalled()
  })

  test('uploadToS3 > handle sign error', async () => {
    vi.spyOn(configMock, 'modelMirror', 'get').mockReturnValue({
      enabled: true,
      export: {
        maxSize: 10,
        kmsSignature: {
          enabled: true,
        },
      },
    })
    kmsMocks.sign.mockRejectedValueOnce('Error')

    await uploadToS3('', {} as unknown as Readable, {} as any)

    expect(logMock.error).toBeCalledWith(expect.objectContaining({}), 'Error generating signature for export.')
    expect(s3Mocks.getObjectStream).toHaveBeenCalled()
  })

  test('uploadToS3 > handle getObjectFromTemporaryS3Location error', async () => {
    vi.spyOn(configMock, 'modelMirror', 'get').mockReturnValue({
      enabled: true,
      export: {
        maxSize: 10,
        kmsSignature: {
          enabled: true,
        },
      },
    })
    s3Mocks.getObjectStream.mockRejectedValueOnce('Error')

    await uploadToS3('', {} as unknown as Readable, {} as any)

    expect(logMock.error).toMatchSnapshot()
    expect(s3Mocks.getObjectStream).toHaveBeenCalled()
  })

  test('uploadToS3 > Handle error when kms enabled', async () => {
    vi.spyOn(configMock, 'modelMirror', 'get').mockReturnValue({
      enabled: true,
      export: {
        maxSize: 10,
        kmsSignature: {
          enabled: true,
        },
      },
    })
    s3Mocks.putObjectStream.mockRejectedValueOnce('Error')

    await uploadToS3('', {} as unknown as Readable, {} as any)

    expect(s3Mocks.putObjectStream).toHaveBeenCalledTimes(1)
    expect(logMock.error).toHaveBeenCalled()
  })

  test('uploadToS3 > Handle error when kms not enabled', async () => {
    vi.spyOn(configMock, 'modelMirror', 'get').mockReturnValue({
      enabled: true,
      export: {
        maxSize: 10,
        kmsSignature: {
          enabled: false,
        },
      },
    })
    s3Mocks.putObjectStream.mockRejectedValueOnce('Error')

    await uploadToS3('', {} as unknown as Readable, {} as any)

    expect(s3Mocks.putObjectStream).toHaveBeenCalledTimes(1)
    expect(logMock.error).toHaveBeenCalled()
  })

  test('getObjectFromExportS3Location > success', async () => {
    await getObjectFromExportS3Location('', {} as any)

    expect(s3Mocks.getObjectStream).toHaveBeenCalled()
    expect(logMock.debug).toHaveBeenCalled()
    expect(logMock.error).not.toHaveBeenCalled()
  })

  test('getObjectFromExportS3Location > throw error', async () => {
    s3Mocks.getObjectStream.mockRejectedValueOnce('Error')

    const promise = getObjectFromExportS3Location('', {} as any)

    await expect(promise).rejects.toThrowError('Error')
    expect(logMock.error).toHaveBeenCalled()
  })
})
