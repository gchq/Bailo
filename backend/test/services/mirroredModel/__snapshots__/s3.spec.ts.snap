// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`services > mirroredModel > s3 > uploadToS3 > handle getObjectFromTemporaryS3Location error 1`] = `
[MockFunction spy] {
  "calls": [
    [
      {
        "error": [Bailo Error: Failed to retrieve stream from temporary S3 location.],
      },
      "Failed to upload export to export location with signatures.",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
